rust/kernel/lib.rs

```diff
@@ -28,12 +28,20 @@
 #![feature(ptr_metadata)]
 #![feature(receiver_trait)]
 #![feature(unsize)]
+// TODO resolve discussion
+#![feature(new_uninit)]
+#![feature(type_alias_impl_trait)]
```

Comment:
I'm less sure about this one. As I understand it there are still outstanding issues (ICE, unsoundness, unexpected behavior) with it that are being fixed: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AF-type_alias_impl_trait

Reply:
yeah, sadly I cannot really work around this one... I need it for the `LockFactory` trait. Maybe it can be done in a different way, but I cannot name a type that implements `PinInit`/`Init`, because it contains a closure...

Reply:
What about writing a custom type implementing `PinInit` rather than `PinInitClosure` with a closure?

Reply:
You are talking about `Mutex::new`, right? You could do the following I think:

```rust
struct RawMutexPinInit {
    name: &'static CStr,
    key: &'static LockClassKey,
}

unsafe impl<T, E> PinInit<Opaque<bindings::mutex>, E> for RawMutexPinInit
{
    unsafe fn __pinned_init(self, slot: *mut Opaque<bindings::mutex>) -> Result<(), E> {
        bindings::__mutex_init(Opaque::raw_get(slot), self.name.as_char_ptr(), self.key.get());
        Ok(())
    }
}

pub const fn new(t: T, name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
    pin_init!(Self {
        mutex: RawMutexPinInit { name, key },
        data: UnsafeCell::new(t),
        _pin: PhantomPinned,
    })
}
```

---

After writing this comment I realized you meant the `pin_init!` call, not `init_mutex()` used inside `pin_init!`. In any case you could do the same transformation for the `pin_init!` I think after macro expanding it, or introduce a `pin_init!` alternative which allows you to explictly declare the `PinInit` implementor name combined with all used upvars. Something like

```rust
pin_init_concrete!(
    // Let pin_init_concrete define a MutexPinInit type which implements PinInit
    MutexPinInit<T> { t: T, name: &'static CStr, key: &'static LockClassKey };
    Self {
        mutex: RawMutexPinInit { name: self.name, key: self.key },
        data: UnsafeCell::new(self.t),
        _pin: PhantomPinned,
    }
)
```

It is harder to write though.

Reply:
How would this work for `SeqLock` though?
```rust
    pub const fn new(
        data: L::Inner,
        name: &'static CStr,
        key1: &'static LockClassKey,
        key2: &'static LockClassKey,
    ) -> impl PinInit<Self, L::Error>
    where
        L: LockFactory<LockedType<L::Inner> = L>,
        L::Inner: Sized,
        <L as LockFactory>::Error: From<core::convert::Infallible>,
    {
        fn init_count(
            name: &'static CStr,
            key2: &'static LockClassKey,
        ) -> impl PinInit<Opaque<bindings::seqcount>> {
            let init = move |place: *mut Opaque<bindings::seqcount>| {
                unsafe {
                    bindings::__seqcount_init(
                        Opaque::raw_get(place),
                        name.as_char_ptr(),
                        key2.get(),
                    )
                };
                Ok(())
            };
            unsafe { PinInitClosure::from_closure(init) }
        }
        pin_init!(Self {
            _p: PhantomPinned,
            count: init_count(name, key2),
            write_lock: L::new_lock(data, name, key1),
        })
    }
```

As it needs to be generic, the macro you are proposing seems *very* complicated...

Reply:
If we had syn it wouldn't be so hard to parse :P On a more serious note, I agree, it is not ideal. I'm just worried people will hit the bugs in type_alias_impl_trait.

Reply:
For sure! I have not encountered a single ICE, so we might only use what is currently implemented correctly...

Reply:
But I think we could use your approach for the types that implement `LockFactory` and then we would not need this feature. I will try and implement it after I fixed some of the other things.